File: Odysseus and the Bag of Winds/d2.py
import sys
from typing import List

def solve_test_case() -> int:
    line = sys.stdin.readline().split()
    M, A = int(line[0]), int(line[1])
    
    cost: List[List[int]] = []
    for i in range(M + 1):
        row = list(map(int, sys.stdin.readline().split()))
        cost.append(row)
    
    INF = 10 ** 15
    full_mask = (1 << M) - 1
    
    dp_no_bag = [[INF] * (M + 1) for _ in range(full_mask + 1)]
    dp_with_bag = [[INF] * (M + 1) for _ in range(full_mask + 1)]
    
    dp_no_bag[1][1] = 0
    
    for mask in range(1, full_mask + 1):
        for current in range(1, M + 1):
            if not ((mask >> (current - 1)) & 1):
                continue
            
            if dp_no_bag[mask][current] < INF:
                for next_node in range(1, M + 1):
                    if (mask >> (next_node - 1)) & 1:
                        continue
                    
                    new_mask = mask | (1 << (next_node - 1))
                    
                    if current == A:
                        cost_with_bag = dp_no_bag[mask][current] + cost[current][next_node] // 2
                        dp_with_bag[new_mask][next_node] = min(
                            dp_with_bag[new_mask][next_node], 
                            cost_with_bag
                        )
                    
                    cost_without_bag = dp_no_bag[mask][current] + cost[current][next_node]
                    dp_no_bag[new_mask][next_node] = min(
                        dp_no_bag[new_mask][next_node],
                        cost_without_bag
                    )
            
            if dp_with_bag[mask][current] < INF:
                for next_node in range(1, M + 1):
                    if (mask >> (next_node - 1)) & 1:
                        continue
                    
                    new_mask = mask | (1 << (next_node - 1))
                    new_cost = dp_with_bag[mask][current] + cost[current][next_node]
                    dp_with_bag[new_mask][next_node] = min(
                        dp_with_bag[new_mask][next_node],
                        new_cost
                    )
    
    return dp_with_bag[full_mask][M]

def main():
    T = int(sys.stdin.readline())
    
    results = []
    for _ in range(T):
        result = solve_test_case()
        results.append(str(result))
    
    print('\n'.join(results))

if __name__ == "__main__":
    main()

File: Odysseus and the Bag of Winds/d2.cpp
#include <iostream>
#include <vector>
#include <algorithm>
using namespace std;

const int INF = 1e9 + 7;

void solve() {
    int M, A;
    cin >> M >> A;
    
    vector<vector<int>> cost(M + 1, vector<int>(M + 1));
    for (int i = 1; i <= M; i++) {
        for (int j = 1; j <= M; j++) {
            cin >> cost[i][j];
        }
    }
    
    int totalMask = (1 << M) - 1;
    vector<vector<int>> dpNoBag(totalMask + 1, vector<int>(M + 1, INF));
    vector<vector<int>> dpWithBag(totalMask + 1, vector<int>(M + 1, INF));
    
    dpNoBag[1][1] = 0;
    dpWithBag[1][1] = INF;
    
    for (int mask = 1; mask <= totalMask; mask++) {
        for (int current = 1; current <= M; current++) {
            if (!((mask >> (current - 1)) & 1)) continue;
            
            if (dpNoBag[mask][current] < INF) {
                for (int next = 1; next <= M; next++) {
                    if ((mask >> (next - 1)) & 1) continue;
                    
                    int newMask = mask | (1 << (next - 1));
                    
                    if (current == A) {
                        int costWithBag = dpNoBag[mask][current] + cost[current][next] / 2;
                        dpWithBag[newMask][next] = min(dpWithBag[newMask][next], costWithBag);
                    }
                    
                    int costWithoutBag = dpNoBag[mask][current] + cost[current][next];
                    dpNoBag[newMask][next] = min(dpNoBag[newMask][next], costWithoutBag);
                }
            }
            
            if (dpWithBag[mask][current] < INF) {
                for (int next = 1; next <= M; next++) {
                    if ((mask >> (next - 1)) & 1) continue;
                    
                    int newMask = mask | (1 << (next - 1));
                    int newCost = dpWithBag[mask][current] + cost[current][next];
                    dpWithBag[newMask][next] = min(dpWithBag[newMask][next], newCost);
                }
            }
        }
    }
    
    cout << dpWithBag[totalMask][M] << endl;
}

int main() {
    ios_base::sync_with_stdio(false);
    cin.tie(nullptr);
    
    int T;
    cin >> T;
    
    while (T--) {
        solve();
    }
    
    return 0;
}